src/query.jl

```diff
@@ -282,3 +282,129 @@ function reducedim_vec(f, x::IndexedTable, dims)
+    end
+    if length(idx) == length(iterdims)
+        idx[end] = vcat(idx[end]) # make last element a vector
+                                  # so that we don't do scalar indexing
+    end
+    y = f(x[idx...]) # Apply on first slice
+
+    if isa(y, IndexedTable)
+        # this means we need to concatenate outputs into a big IndexedTable
+        ns = vcat(dimlabels(x)[iterdims], dimlabels(y))
+        if !all(x->isa(x, Symbol), ns)
+            ns = nothing
+        else
+            ns = dedup_names(ns)
+        end
+        n = length(y)
+        index_first = similar(iter, n)
+        for j=1:n
+            @inbounds index_first[j] = iter[1]
+        end
+        index = Columns(index_first.columns..., astuple(y.index.columns)...; names=ns)
+        data = copy(y.data)
+        output = IndexedTable(index, data)
+        _mapslices_itable!(f, output, x, iter, iterdims, 1)
+    else
+        ns = dimlabels(x)[iterdims]
+        if !all(x->isa(x, Symbol), ns)
+            ns = nothing
+        end
+        index = Columns(iter[1:1].columns...; names=ns)
+        if name === nothing
+            output = IndexedTable(index, [y])
+        else
+            output = IndexedTable(index, Columns([y], names=[name]))
+        end
+        _mapslices_scalar!(f, output, x, iter, iterdims, 2, name!==nothing ? x->(x,) : identity)
+    end
+end
+
+function _mapslices_scalar!(f, output, x, iter, iterdims, start, coerce)
+    idx = Any[Colon() for v in x.index.columns]
+
+    for i = start:length(iter)
+        if i != 1 && roweq(iter, i-1, i) # We've already visited this slice
+            continue
+        end
+        for j in 1:length(iterdims)
+            d = iterdims[j]
+            idx[d] = iter[i][j]
+        end
+        y = f(x[idx...])
+        push!(output.index, iter[i])
+        push!(output.data, coerce(y))
+    end
+    output
+end
+
+function _mapslices_itable!(f, output, x, iter, iterdims, start)
+    idx = Any[Colon() for v in x.index.columns]
+    I = output.index
+    D = output.data
+    srcdims = length(iterdims)
+
+    l = length(output)
+    for i = start:length(iter)
+        if i != 1 && roweq(iter, i-1, i) # We've already visited this slice
+            continue
+        end
+        for j in 1:length(iterdims)
+            d = iterdims[j]
+            idx[d] = iter[i][j]
+        end
+        y = f(x[idx...])
+        n = length(y)
+        resize!(I, l+n)
+        resize!(D, l+n)
+
+        for k = 1:srcdims
+            ldestcol = I.columns[k]
+            srccol = iter.columns[k]
+            @inbounds for j=l+1:l+n
+                ldestcol[j] = srccol[i]
+            end
+        end
+
+        for k = (srcdims+1):srcdims+length(y.index.columns)
+            ldestcol = I.columns[k]
+            srccol = y.index.columns[k-srcdims]
+            @inbounds for j=l+1:l+n
+                ldestcol[j] = srccol[j-l]
+            end
+        end
+
+        @inbounds for j=l+1:l+n
+            D[j] = y.data[j-l]
```

Comment:
Replaced... I was trying to avoid creating new `Columns` for every slice just to append... worked around that by splitting `I` into 2 sets `Columns` in the beginning. I also added an `append_n!` to utils.jl, this is 3-4x faster than using `append!(x, Iterators.repeated(v,n))`